### PR TITLE
React section: Adds Class-based components lesson

### PR DESCRIPTION
**Pull request by Lofty-Brambles**
Complete the following REQUIRED checkboxes:

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`

Complete the following checkboxes ONLY IF they are applicable to your PR. You can complete them later if they are not currently applicable:
-   [x] I have previewed all lesson files included in this PR with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure the Markdown content is formatted correctly
-   [x] I have ensured all lesson files included in this PR follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

<hr>

**1. Because:**

Adds the lesson based on [top-meta #230](https://github.com/TheOdinProject/top-meta/issues/230)

**2. This PR:**

- Adds a class-based components lesson.

**3. Additional Information:**

N/A. Waiting on feedback, and merging of the PR, adding the example for the codesandbox in the react-examples repository.
Just as an example, I have a demo one provided, which will be replaced when the codesandbox will get added, but this is in place for a better showcase.

### CODE REVIEW

File: react/class_components/class_based_components.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        {todos.map((todo) => (
+          <li key={todo}>{todo}</li>
+        ))}
+      </ul>
+    </section>
+  );
+};
+
+export default FunctionalInput;
+~~~
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:74]: We should remove comments in this code example. By this point, learners should be well versed in functional components and how state works.

File: react/class_components/class_based_components.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+That was a solid chunk of code. Take a while, sip some water and read it a couple of times.
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:76]: <details>
<summary></summary>

image

File: react/class_components/class_based_components.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+This section contains a general overview of topics that you will learn in this lesson.
+
+- A note on the relevance of class components.
+- Covers the structure of a class component, the uses of the constructor and render method.
+- Discusses the ways to use props and state in class components.
+- Highlights the uses of `this` in class components.
+
+### The What's And Why's
+
+In your previous lessons, you have already been introduced to functional components, and the basic patterns in which components get written now a days. However, React components did not look this way since React was introduced.
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:16]:
```suggestion
In your previous lessons, you have already been introduced to functional components, and the basic patterns in which components get written now a days. However, React components did not look this way when React was introduced.
```

File: react/class_components/class_based_components.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+If you look into any older React codebase, you'll notice a lot of classes. These are known as class-based components. Prior to February 2019, functional components were also called state-less, as there was no way to manage state in them. This was changed when hooks were introduced, leading to less verbose and 'neater' components.
+
+In your career, chances are, you will be dealing with legacy code, so there will be days where you would be dealing with class components. To meet this issue, we take a peek into the works and intricacies of a class-based component, and how they work.
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:20]: Wording suggestion:

```suggestion
In your career, chances are, you will be dealing with legacy code, so there will be days where you would be dealing with class components. Let's peek into the intricacies of a class-based component, and how they work.
```

File: react/class_components/class_based_components.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+All the components by far, have been functional in style and syntax. This is common now, but you will see a different `class` based syntax too. In this lesson, we explore how a class-based component is written and the significance of the different elements seen in one.
+
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- A note on the relevance of class components.
+- Covers the structure of a class component, the uses of the constructor and render method.
+- Discusses the ways to use props and state in class components.
+- Highlights the uses of `this` in class components.
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:12]: We can simplify the Lesson Overview:

```suggestion

- Learn the structure of a class component and how they're written.
- How to use props and state in class components.
- Highlight the uses of `this` in class components.
```

File: react/class_components/class_based_components.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+<div class="lesson-content__panel" markdown="1">
+
+For the purposes of this assignment, we take the simple class-based component that we built. There is an interactive example provided at the end, where you can work on.
+
+1.  Implement a delete button for each task. The delete button should remove that specific task from the state array, thus deleting the task itself! Styling isn't a priority at this moment, but the button tag should be styled by default.
+
+2.  Implement a new class component, `Count` that displays the count of the number of todos, at any given time.
+
+3.  Implement an edit button for each task. It should replace the todo with an input field, and change the button itself to 'Resubmit', so the edits can be saved. This is a comparatively harder task, kudos for when you finish it!
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:277]: I like the assignment section. Hands on practice is invaluable!

**Review by 01zulfi — CHANGES_REQUESTED** [review]
Amazing work you've put in here 🔥 🔥 🔥 

Left some comments below.

**Review by 01zulfi — CHANGES_REQUESTED** [review]
I'll take a look at the code examples in a bit more detail once I get a chance to review the codesandbox PR (promise I'll get to it soon 😄). Left a couple of comments below, the lesson is looking solid otherwise.

File: react/class_components/class_based_components.md [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- Learn the structure of a class component and how they are written.
+- How to use props and state in class components.
+- Highlight the uses of `this` in class components.
+
+### The What's And Why's
```
[the remark this answers is not in the document]
Comment (Lofty-Brambles) [reply react/class_components/class_based_components.md:13]: Grammar gosh.. [Here's what I found](https://english.stackexchange.com/questions/146348/the-whys-and-the-hows), so I'm simply dropping them. Thanks for that, learnt something today-

**Comment by Lofty-Brambles** [comment]
Addressed the two issues @01zulfi

File: react/class_components/class_based_components.md [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+2.  Implement a new class component, `Count` that displays the count of the number of todos, at any given time.
+
+3.  Implement an edit button for each task. It should replace the todo with an input field, and change the button itself to 'Resubmit', so the edits can be saved. This is a comparatively harder task, kudos for when you finish it!
+
+<iframe src='https://codesandbox.io/embed/component-comparisions-difgyb?fontsize=14&hidenavigation=1&theme=dark'
+        style='width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;'
+        title='Component comparisions'
+        allow='accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking'
+        sandbox='allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts'
+></iframe>
```
Comment (01zulfi) [inline react/class_components/class_based_components.md:272]: I've created a sandbox on TOP's account, we should change to that link instead: 
```
<iframe src="https://codesandbox.io/embed/github/TheOdinProject/react-examples/tree/main/class-component-example?fontsize=14&hidenavigation=1&theme=dark"
     style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;"
     title="react-example"
     allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
     sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
   ></iframe>
```

**Review by 01zulfi — CHANGES_REQUESTED** [review]

**Comment by Lofty-Brambles** [comment]
@01zulfi - It should be added! Could you please ping me when you add another comment next time?

**Review by 01zulfi — APPROVED** [review]

**Pull request merged**